kani-compiler/src/kani_middle/points_to/points_to_graph.rs

```diff
@@ -0,0 +1,216 @@
+    Alloc(DefId, Location),
+    Place(Place<'tcx>),
+}
+
+/// A node tagged with a DefId, to differentiate between places across different functions.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub enum GlobalMemLoc<'tcx> {
+    Local(DefId, LocalMemLoc<'tcx>),
+    Global(DefId),
+}
+
+impl<'tcx> GlobalMemLoc<'tcx> {
+    /// Returns DefId of the memory location.
+    pub fn def_id(&self) -> DefId {
+        match self {
+            GlobalMemLoc::Local(def_id, _) | GlobalMemLoc::Global(def_id) => *def_id,
+        }
+    }
+
+    /// Returns LocalMemLoc of the memory location if available.
+    pub fn maybe_local_mem_loc(&self) -> Option<LocalMemLoc<'tcx>> {
+        match self {
+            GlobalMemLoc::Local(_, mem_loc) => Some(*mem_loc),
+            GlobalMemLoc::Global(_) => None,
+        }
+    }
+}
+
+impl<'tcx> From<Place<'tcx>> for LocalMemLoc<'tcx> {
+    fn from(value: Place<'tcx>) -> Self {
+        LocalMemLoc::Place(value)
+    }
+}
+
+impl<'tcx> LocalMemLoc<'tcx> {
+    /// Register a new heap allocation site.
+    pub fn new_alloc(def_id: DefId, location: Location) -> Self {
+        LocalMemLoc::Alloc(def_id, location)
+    }
+
+    /// Tag the node with a DefId.
+    pub fn with_def_id(&self, def_id: DefId) -> GlobalMemLoc<'tcx> {
+        GlobalMemLoc::Local(def_id, *self)
+    }
+}
+
+/// Graph data structure that stores the current results of the point-to analysis. The graph is
+/// directed, so having an edge between two places means that one is pointing to the other. For
+/// example, `a = &b` would translate to `a --> b` and `a = b` to `a --> {all pointees of b}`.
+///
+/// Note that the aliasing is stored between places with no projections, which is sound but can be
+/// imprecise. I.e., if two places have an edge in the graph, could mean that some scalar sub-places
+/// (e.g. _1.0) of the places alias, too, but not the deref ones.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PointsToGraph<'tcx> {
+    /// A hash map of node --> {nodes} edges.
+    edges: HashMap<GlobalMemLoc<'tcx>, HashSet<GlobalMemLoc<'tcx>>>,
+}
+
+impl<'tcx> PointsToGraph<'tcx> {
+    pub fn empty() -> Self {
+        Self { edges: HashMap::new() }
+    }
+
+    /// Collect all nodes which have incoming edges from `nodes`.
+    pub fn follow(&self, nodes: &HashSet<GlobalMemLoc<'tcx>>) -> HashSet<GlobalMemLoc<'tcx>> {
+        nodes.iter().flat_map(|node| self.edges.get(node).cloned().unwrap_or_default()).collect()
+    }
+
+    /// For each node in `from`, add an edge to each node in `to`.
+    pub fn extend(&mut self, from: &HashSet<GlobalMemLoc<'tcx>>, to: &HashSet<GlobalMemLoc<'tcx>>) {
+        for node in from.iter() {
+            let node_pointees = self.edges.entry(*node).or_default();
+            node_pointees.extend(to.iter());
+        }
+    }
+
+    /// Collect all scalar places to which a given place can alias. This is needed to resolve all
```

Comment:
I couldn't find the definition of `scalar places`. So I don't quite understand what you mean here.

Reply:
I removes all references to "scalar places", hopefully it is more clear now